File: ios/BetterLise/Features/Agenda/AgendaViewModel.swift
import Foundation
import Observation

@MainActor
@Observable
final class AgendaViewModel {
    /// School weeks reachable by swiping on each side of the current week.
    static let weeksAround = 26

    private(set) var state: Loadable<[CalendarEvent]> = .idle
    private(set) var eventsByDay: [Date: [CalendarEvent]] = [:]

    let days: [Date]
    let today: Date
    let todayIndex: Int

    /// Week shown by the pager.
    private(set) var visibleWeek: Int

    /// Programmatic scroll the pager must perform ("Aujourd'hui").
    private(set) var pagerRequest: ScrollRequest?

    private let session: SessionStore
    private let settings: SettingsStore
    private let cache: ResponseCache

    init(session: SessionStore, settings: SettingsStore, cache: ResponseCache, now: () -> Date = Date.init) {
        self.session = session
        self.settings = settings
        self.cache = cache
        today = now()
        days = AgendaLayout.schoolDays(around: today, weeksBefore: Self.weeksAround, weeksAfter: Self.weeksAround)
        todayIndex = AgendaLayout.initialIndex(in: days, today: today)
        visibleWeek = todayIndex / 5
    }

    var weekCount: Int { days.count / 5 }
    /// Week of today, or the upcoming one on weekends.
    var currentWeek: Int { todayIndex / 5 }
    var isShowingCurrentWeek: Bool { visibleWeek == currentWeek }

    /// Monday of the week currently shown, used for the title.
    var visibleWeekStart: Date { days[min(visibleWeek * 5, days.count - 1)] }

    /// The user swiped the pager.
    func pagerDidScroll(toWeek week: Int) {
        guard (0..<weekCount).contains(week), week != visibleWeek else { return }
        visibleWeek = week
    }

    func goToToday() {
        visibleWeek = currentWeek
        pagerRequest = ScrollRequest(target: currentWeek)
    }

    func days(inWeek week: Int) -> ArraySlice<Date> {
        let start = week * 5
        return days[start..<min(start + 5, days.count)]
    }

    func events(on day: Date) -> [CalendarEvent] {
        eventsByDay[Calendar.paris.startOfDay(for: day)] ?? []
    }

    func hasAllDayEvents(inWeek week: Int) -> Bool {
        days(inWeek: week).contains { day in events(on: day).contains(where: \.isAllDay) }
    }

    private var cacheKey: String { "agenda-\(settings.liseId)-\(settings.campus.rawValue)-\(settings.showRU)" }

    func load() async {
        guard settings.hasValidLiseId else {
            apply(.idle)
            return
        }
        let cached = cache.load([CalendarEvent].self, key: cacheKey) ?? state.value
        apply(.loading(cached: cached))
        do {
            let response = try await session.sendPublic(
                Endpoints.agenda(liseId: settings.liseId, tbk: settings.campus.rawValue, includeRU: settings.showRU)
            )
            cache.save(response.events, key: cacheKey)
            apply(.loaded(response.events))
        } catch {
            apply(.failed(message: error.localizedDescription, cached: cached))
        }
    }

    private func apply(_ newState: Loadable<[CalendarEvent]>) {
        if newState.value != state.value {
            eventsByDay = AgendaLayout.groupByDay(newState.value ?? [])
        }
        state = newState
    }
}

/// A one-shot scroll command. The UUID makes two requests for the same target distinct.
struct ScrollRequest: Equatable {
    let target: Int
    private let id = UUID()

    init(target: Int) {
        self.target = target
    }
}

File: android/app/src/main/java/com/betterlise/app/ui/agenda/AgendaViewModel.kt
package com.betterlise.app.ui.agenda

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.betterlise.app.data.api.CalendarEvent
import com.betterlise.app.data.api.Endpoints
import com.betterlise.app.data.auth.SessionRepository
import com.betterlise.app.data.cache.ResponseCache
import com.betterlise.app.data.settings.SettingsRepository
import com.betterlise.app.data.settings.UserSettings
import com.betterlise.app.domain.AgendaLayout
import com.betterlise.app.domain.PARIS
import com.betterlise.app.ui.components.Loadable
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Job
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import kotlinx.serialization.builtins.ListSerializer
import java.time.LocalDate

/** A one-shot scroll command; [nonce] makes two requests for the same target distinct. */
data class ScrollRequest(val target: Int, val nonce: Long)

data class AgendaUiState(
    val settings: UserSettings = UserSettings(),
    val events: Loadable<List<CalendarEvent>> = Loadable.Idle,
    val eventsByDay: Map<LocalDate, List<CalendarEvent>> = emptyMap(),
    val days: List<LocalDate>,
    val today: LocalDate,
    val todayIndex: Int,
    val visibleWeek: Int = todayIndex / 5,
    /** Programmatic scroll the week pager must perform ("Aujourd'hui"). */
    val pagerRequest: ScrollRequest? = null,
) {
    val weekCount: Int get() = days.size / 5
    /** Week of today, or the upcoming one on weekends. */
    val currentWeek: Int get() = todayIndex / 5
    val isShowingCurrentWeek: Boolean get() = visibleWeek == currentWeek
    val visibleWeekStart: LocalDate get() = days[(visibleWeek * 5).coerceAtMost(days.lastIndex)]

    fun eventsOn(day: LocalDate): List<CalendarEvent> = eventsByDay[day].orEmpty()
    fun indicesInWeek(week: Int): IntRange = (week * 5) until minOf(week * 5 + 5, days.size)
    fun hasAllDayEvents(week: Int): Boolean = indicesInWeek(week).any { index -> eventsOn(days[index]).any { it.isAllDay } }
}

class AgendaViewModel(
    private val session: SessionRepository,
    private val settingsRepository: SettingsRepository,
    private val cache: ResponseCache,
    today: () -> LocalDate = { LocalDate.now(PARIS) },
) : ViewModel() {
    private val _state: MutableStateFlow<AgendaUiState>
    val state: StateFlow<AgendaUiState>
    private var loadJob: Job? = null
    private var requestCounter = 0L

    init {
        val now = today()
        val days = AgendaLayout.schoolDays(now, WEEKS_AROUND, WEEKS_AROUND)
        _state = MutableStateFlow(AgendaUiState(days = days, today = now, todayIndex = AgendaLayout.initialIndex(days, now)))
        state = _state.asStateFlow()

        viewModelScope.launch {
            settingsRepository.settings
                .distinctUntilChanged { old, new ->
                    old.liseId == new.liseId && old.campus == new.campus && old.showRu == new.showRu
                }
                .collect { settings ->
                    _state.update { it.copy(settings = settings) }
                    refresh()
                }
        }
    }

    /** The user swiped the week pager. */
    fun pagerDidScroll(week: Int) = _state.update { current ->
        if (week !in 0 until current.weekCount || week == current.visibleWeek) current else current.copy(visibleWeek = week)
    }

    fun goToToday() = _state.update { current ->
        current.copy(visibleWeek = current.currentWeek, pagerRequest = nextRequest(current.currentWeek))
    }

    private fun nextRequest(target: Int) = ScrollRequest(target, ++requestCounter)

    fun refresh() {
        val settings = _state.value.settings
        if (!settings.hasValidLiseId) {
            _state.update { it.copy(events = Loadable.Idle, eventsByDay = emptyMap()) }
            return
        }
        loadJob?.cancel()
        loadJob = viewModelScope.launch {
            val key = "agenda-${settings.liseId}-${settings.campus.id}-${settings.showRu}"
            val cached = cache.load(key, EVENTS) ?: _state.value.events.value
            _state.update { it.withEvents(Loadable.Loading(cached)) }
            try {
                val response = session.sendPublic(Endpoints.agenda(settings.liseId, settings.campus.id, settings.showRu))
                cache.save(key, EVENTS, response.events)
                _state.update { it.withEvents(Loadable.Loaded(response.events)) }
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                _state.update { it.withEvents(Loadable.Failed(e.message ?: "Erreur", cached)) }
            }
        }
    }

    private fun AgendaUiState.withEvents(events: Loadable<List<CalendarEvent>>): AgendaUiState =
        if (events.value == this.events.value) {
            copy(events = events)
        } else {
            copy(events = events, eventsByDay = AgendaLayout.groupByDay(events.value.orEmpty()))
        }

    companion object {
        /** School weeks reachable by swiping on each side of the current week. */
        const val WEEKS_AROUND = 26
        private val EVENTS = ListSerializer(CalendarEvent.serializer())
    }
}
